docs: add more orchestrator documentation (#346)
Substra/substra-documentation: docs/source/documentation/orchestrator/index.rst

Orchestrator
************

Performing a Federated Learning experiment implies a lot of different compute tasks: local training, aggregation, testing on different organizations, etc. The role of the orchestrator is to distribute ML tasks among organizations, while ensuring complete traceability of operations.

The orchestrator registers the status of tasks; when a task is done, it evaluates if some tasks in ``Waiting`` are now unblocked, and if it's the case, the status of those tasks is changed to ``To do``. The new status is sent to all the backends, who store the new tasks ``To do`` in the task queue (Celery). Then, the task queue will assign the task to one of the workers (if multiple) and handle retries if needed.

Centralized vs. decentralized orchestration
===========================================

> Substra offers two types of orchestration: distributed and centralized.

Reviewer: ```suggestion
Substra offers two types of orchestration: **distributed** and **centralized**.
```

> .. image:: /static/schemes/distributed-vs-centralized-orc.png

Reviewer: The scheme background is blank, you need to make it transparent (or it'll be weird on the doc page)

The distributed orchestration is based on a private blockchain using Hyperledger Fabric, while the centralized orchestration is hosted by a central Postgres database.

In both cases, the orchestration stores only non-sensitive metadata of the Substra assets, makes it possible to verify the integrity of the assets and ensures that the permissions on the assets are respected.

Distributed orchestration enables trustless verification of the integrity of assets (functions, model, data), but it requires connections between organizations, and introduces a network overhead. It's not possible to upgrade a Substra network when using distributed orchestration.

On the other hand, centralized orchestration requires trust in the central server, but it is more efficient, faster and easier to deploy and maintain.

> As long as you trust whomever is operating the orchestrator DB not to tamper with it, both modes offer the same level of guarantees. The decentralized mode has nice theoretical guarantees, but the network overhead is very significant, and has other a lot of operational drawbacks. That is why, today the vast majority (if not all) of the Substra deployments are using the centralized orchestration system as it is easier to operate. However the distributed orchestration is still maintained.

Reviewer: ```suggestion
As long as you trust whomever is operating the orchestrator DB not to tamper with it, both modes offer the same level of guarantees. The decentralized mode has nice theoretical guarantees, but the network overhead is very significant, and has a lot of operational drawbacks. That is why, the vast majority (if not all) of the Substra deployments are using the centralized orchestration system as it is easier to operate and faster. However, the distributed orchestration is still maintained.
```

Orchestrator
************

Performing a Federated Learning experiment implies a lot of different compute tasks: local training, aggregation, testing on different organizations, etc. The role of the orchestrator is to distribute ML tasks among organizations, while ensuring complete traceability of operations.

> The orchestrator registers the status of tasks; when a task is done, it evaluates if some tasks in ``Waiting`` are now unblocked, and if it's the case, the status of those tasks is changed to ``To do``. The new status is sent to all the backends, who store the new tasks ``To do`` in the task queue (Celery). Then, the task queue will assign the task to one of the workers (if multiple) and handle retries if needed.

Reviewer 2: ```suggestion
The orchestrator registers the status of tasks; when a task is done (status ``Done``), it evaluates if some remaining tasks (status ``Waiting``) are now unblocked, and if it's the case, the status of those tasks is changed to ``To do``. The new status is sent to all the backends, who store the new tasks ``To do`` in the task queue (Celery). Then, the task queue will assign the task to one of the workers (if multiple) and handle retries if needed.

In case of failure, it will store fialure reports and  change the status of the faulty task to ``Failed``.
In case of manual cancellation, it will change the status of the  the tasks to `Cancelled` on different backends.
```

Centralized vs. decentralized orchestration
===========================================

Substra offers two types of orchestration: **distributed** and **centralized**.

.. image:: /static/schemes/distributed-vs-centralized-orc.svg

The distributed orchestration is based on a private blockchain using Hyperledger Fabric, while the centralized orchestration is hosted by a central Postgres database.

> In both cases, the orchestration stores only non-sensitive metadata of the Substra assets, makes it possible to verify the integrity of the assets and ensures that the permissions on the assets are respected.

Reviewer 2: ```suggestion
In both cases, the orchestration stores only non-sensitive metadata of the Substra assets, making it possible to verify the integrity of the assets and ensures that the permissions on the assets are respected.
```

Distributed orchestration enables trustless verification of the integrity of assets (functions, model, data), but it requires connections between organizations, and introduces a network overhead. It's not possible to upgrade a Substra network when using distributed orchestration.

> On the other hand, centralized orchestration requires trust in the central server, but it is more efficient, faster and easier to deploy and maintain.

Reviewer 2: Not sure that the meaning of *efficient* is straightforward in this case

Author: I removed it.

> As long as you trust whomever is operating the orchestrator DB not to tamper with it, both modes offer the same level of guarantees. The decentralized mode has nice theoretical guarantees, but the network overhead is very significant, and has a lot of operational drawbacks. That is why, the vast majority (if not all) of the Substra deployments are using the centralized orchestration system as it is easier to operate and faster. However, the distributed orchestration is still maintained.

Reviewer 2: > That is why, the vast majority (if not all) of the Substra deployments

Should we emphazise that is the current situation but historically it may have been different ? (i.e. *That is why, the vast majority (if not all) of the **current** Substra deployments*)

Author: agree, added
